kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -395,20 +396,87 @@ private static void validateSchemaEvolution(
     }
   }
 
-  /* Validates a given schema evolution by using field ID
-  as the source of truth for identifying fields */
-  private static void validateSchemaEvolutionById(
-      StructType currentSchema, StructType newSchema, Metadata metadata) {
+  /**
+   * Validates a given schema evolution by using field ID as the source of truth for identifying
+   * fields
+   */
+  private static void validateSchemaEvolutionById(StructType currentSchema, StructType newSchema) {
     Map<Long, StructField> currentFieldsById = fieldsById(currentSchema);
     Map<Long, StructField> updatedFieldsById = fieldsById(newSchema);
     SchemaChanges schemaChanges = computeSchemaChangesById(currentFieldsById, updatedFieldsById);
-
-    // Validate physical name consistency on the updated fields
     validatePhysicalNameConsistency(schemaChanges.updatedFields());
+    validateUpdatedSchemaCompatibility(schemaChanges);
+    // ToDo Potentially validate IcebergCompatV2 nested IDs
+  }
 
-    // ToDo Validate IcebergCompatV2 nested IDs
+  /**
+   * Verifies the following
+   *
+   * <ul>
+   *   <li>no non-nullable fields are added
+   *   <li>no tightening of existing fields
+   *   <li>no type changes are performed
+   */
+  private static void validateUpdatedSchemaCompatibility(SchemaChanges schemaChanges) {
+    for (StructField addedField : schemaChanges.addedFields()) {
+      if (!addedField.isNullable()) {
+        throw new KernelException(
+            String.format("Cannot add non-nullable field %s", addedField.getName()));
+      }
+    }
 
-    // ToDo: Validate schema compatibility
+    for (Tuple2<StructField, StructField> updatedFields : schemaChanges.updatedFields()) {
+      validateNoTypeChange(updatedFields._1, updatedFields._2);
+    }
+  }
+
+  /**
+   * Validate that there was no change in type from existing field from new field, excluding dropped
+   * columns for structs.
+   */
+  private static void validateNoTypeChange(StructField existingField, StructField newField) {
+    if (existingField.isNullable() && !newField.isNullable()) {
+      throw new KernelException(
+          String.format(
+              "Cannot tighten the nullability of existing field %s", existingField.getName()));
+    }
+
+    // Both fields are structs, ensure there's no changes in the individual fields
+    // ToDo: Prevent modifications to map keys where the key is a struct
```

Comment:
I added this ToDo because Delta allows adding/removing struct fields when the struct is a key in a map, however Iceberg doesn't. I think we can discuss separately if we want to be more restrictive here by default or not.